daphne_worker/src/storage_proxy/metrics.rs

```diff
@@ -0,0 +1,148 @@
+
+use std::{collections::HashMap, sync::OnceLock};
+
+use daphne::auth::BearerToken;
+use prometheus::{
+    register_int_counter_vec_with_registry, Encoder, IntCounterVec, Registry, TextEncoder,
+};
+use url::Url;
+use worker::Env;
+
+struct Config {
+    prometheus_server: Url,
+
+    bearer_token: BearerToken,
+}
+
+impl Config {
+    fn from_env(env: &Env) -> Option<&'static Self> {
+        const DAP_METRICS_PUSH_SERVER_URL: &str = "DAP_METRICS_PUSH_SERVER_URL";
+        const DAP_METRICS_PUSH_BEARER_TOKEN: &str = "DAP_METRICS_PUSH_BEARER_TOKEN";
+        static CONFIG: OnceLock<Option<Config>> = OnceLock::new();
+
+        CONFIG
+            .get_or_init(|| {
+                match (
+                    env.var(DAP_METRICS_PUSH_SERVER_URL),
+                    env.var(DAP_METRICS_PUSH_BEARER_TOKEN),
+                ) {
+                    (Ok(server_str), Ok(bearer_token_str)) => {
+                        let prometheus_server = match server_str.to_string().parse() {
+                            Ok(server) => server,
+                            Err(error) => {
+                                tracing::error!(error = ?error, "invalid server url");
+                                return None;
+                            }
+                        };
+                        Some(Self {
+                            prometheus_server,
+                            bearer_token: BearerToken::from(bearer_token_str.to_string()),
+                        })
+                    }
+                    (Err(_), Err(_)) => None,
+                    (Ok(_), Err(error)) => {
+                        tracing::error!(%error, "failed to configure metrics push: missing bearer token");
+                        None
+                    }
+                    (Err(error), Ok(_)) => {
+                        tracing::error!(%error, "failed to configure metrics push: missing server URL");
+                        None
+                    }
+                }
+            })
+            .as_ref()
+    }
+}
+
+pub struct Metrics {
+    config: &'static Config,
+
+    _registry: Registry,
+
+    /// Number of retries done in durable object requests before returning (whether by success
+    /// or failure).
+    durable_request_retry_count: IntCounterVec,
+
+    /// HTTP response status.
+    http_status_code_counter: IntCounterVec,
+}
+
+impl Metrics {
+    pub fn new(host: &str, env: &Env) -> Option<Self> {
+        let config = Config::from_env(env)?;
+
+        let registry = Registry::new_custom(
+            Option::None,
+            Option::Some(HashMap::from([("host".to_string(), host.to_string())])),
+        )
+        .unwrap();
+
+        let durable_request_retry_count = register_int_counter_vec_with_registry!(
+            "durable_request_retry_count",
+            "The number of times a request to durable objects was retried before succeeding or giving up",
+            &["count", "object", "path"],
```

Comment:
Actually, what variants of the "count" label will there be? Don't we just want a counter for each "object"/"path" pair?

Reply:
Updated the description of the metric to hopefully make it more obvious.

`count` it will be a number in the range [-1, 4].
- `-1`: the request failed
- `0`: no retries were done
- `1`..`4`: X retires were done but we succeeded

Reply:
Ahhhh, it keepts track of the number of times it was retired. That works!